app/javascript/hooks/useLazyFetch.ts

```diff
@@ -0,0 +1,190 @@
+import React from "react";
+
+import { assertResponseError, request } from "$app/utils/request";
+
+import { showAlert } from "$app/components/server-components/Alert";
+
+interface UseLazyFetchOptions<T> {
+  url: string;
+  responseParser: (data: unknown) => T;
+  hasMore?: boolean;
+}
+
+interface UseLazyFetchResult<T> {
+  data: T;
+  isLoading: boolean;
+  setIsLoading: (isLoading: boolean) => void;
+  setData: (data: T) => void;
+  fetchData: (queryParams?: QueryParams) => Promise<void>;
+  hasLoaded: boolean;
+  setHasLoaded: (hasLoaded: boolean) => void;
+}
+
+type QueryParams = Record<string, string | number>;
+
+export type Pagination = {
+  count: number;
+  from: number;
+  in: number;
+  last: number;
+  limit: number;
+  next: number | null;
+  offset: number;
+  outset: number;
+  overflow: number;
+  page: number;
+  prev: number | null;
+  to: number;
```

Comment:
Do we use / care about all these? It's a bit of an overwhelming list at first sight